app/tor.js

```diff
@@ -0,0 +1,1253 @@
+        if (chunk[i] === 0x0a) {         // LF
+          // CRLF seen.  Concatenate what we have, minus CRLF.
+          assert(start < end)
+          this._chunks.push(chunk.slice(start, end))
+          this._readlen += end - start
+          this._line(false)
+          start = i + 1
+          end = chunk.length
+        } else {
+          // CR seen, but not LF.  Bad.
+          return this._error(chunk, start, 'stray carriage return')
+        }
+      }
+    }
+    // No CRLF in the chunk.  Store up to CR from chunk and go on.
+    assert(start <= end)
+    assert(this._maxlen === null || this._readlen <= this._maxlen)
+    assert(this._maxlen === null || end - start <= this._maxlen - this._readlen)
+    if (start < end) {
+      this._chunks.push(chunk.slice(start, end))
+      this._readlen += end - start
+    }
+    assert(this._maxlen === null || this._readlen <= this._maxlen)
+  }
+
+  // LineReadable._onEnd()
+  //
+  //    Event handler for end of underlying readable stream.  Reports
+  //    a final line if we have any data, tidies up after ourselves,
+  //    and emits an end event.
+  //
+  _onEnd () {
+    // If there's anything stored, report it.
+    if (this._readlen !== 0) {
+      this._line(false)
+    }
+    // Tidy up after ourselves.
+    this._tidy()
+    // Emit the event.
+    this.emit('end')
+  }
+
+  // LineReadable._line(trunc)
+  //
+  //    Report a line event, possibly truncated, with the chunks of
+  //    data so far accumulated.  Reset the state to the beginning of
+  //    a line.
+  //
+  _line (trunc) {
+    // Compute the line.
+    const line = Buffer.concat(this._chunks, this._readlen)
+    // Reset the state.
+    this._reset()
+    // Emit the event.
+    this.emit('line', line, trunc)
+  }
+
+  // LineReadable._error(chunk, start, msg)
+  //
+  //    Report an error event with the specified message.  Return any
+  //    data -- chunk[start], chunk[start+1], ..., chunk[n-1] -- to
+  //    the stream.  Reset and tidy up, since we're presumed wedged.
+  //
+  _error (chunk, start, msg) {
+    // Add what's left of the current chunk.
+    assert(start < chunk.length)
+    this._chunks.push(chunk.slice(start))
+    this._readlen += chunk.length - start
+    // Restore the chunks we've consumed to the stream.
+    this._readable.unshift(Buffer.concat(this._chunks, this._readlen))
+    // Reset the state.  No need to hang onto garbage we won't use.
+    this._reset()
+    // Tidy up after ourselves: we are wedged and uninterested in
+    // further events.
+    this._tidy()
+    // Emit the event.
+    this.emit('error', new Error(msg))
+  }
+}
+
+module.exports.TorControl = TorControl
+module.exports.TorDaemon = TorDaemon
+module.exports.torControlParseKV = torControlParseKV
+module.exports.torControlParseQuoted = torControlParseQuoted
+module.exports.torrcEscapeBuffer = torrcEscapeBuffer
+module.exports.torrcEscapePath = torrcEscapePath
+module.exports.torrcEscapeString = torrcEscapeString
```

Comment:
please add unit tests for these exported methods (located in test/unit, `npm run unittest` to run

Reply:
done in 428280759b7f9fe963d3bea7a9b7f169311ecf44

Reply:
thanks for adding tests - could you also document the input types of any externally-callable methods/constructors using jsdoc so it's harder to get the types wrong? http://usejsdoc.org/ - see js/lib/request for an example

Reply:
Will do.

Reply:
done in 8f532a9a58155a82d65050fce4d647193d9dc894